### PR TITLE
Improve CA error message and CA server monitoring.

### PR DESCRIPTION
**Pull request by myidpt**

### CODE REVIEW

File: security/pkg/server/ca/server.go [hunk 1]
```diff
@@ -60,30 +60,32 @@ func (s *Server) HandleCSR(ctx context.Context, request *pb.CsrRequest) (*pb.Csr
...
+		log.Warnf("CSR Pem parsing error (error %v)", err)
+		s.monitoring.GetAuthzError("CSR Pem parsing error").Inc()
 		return nil, status.Errorf(codes.InvalidArgument, "CSR parsing error (%v)", err)
 	}
 
 	_, err = util.ExtractIDs(csr.Extensions)
 	if err != nil {
 		log.Warnf("CSR identity extraction error (%v)", err)
-		s.monitoring.IDExtractionError.Inc()
+		s.monitoring.GetAuthzError("CSR identity extraction error").Inc()
```
[the remark this answers is not in the document]
Comment (myidpt) [reply security/pkg/server/ca/server.go]: Good suggestions. I have made a error enum in ca/error.go to solve this problem.

**Comment by myidpt** [comment]
@incfly Yes, we should also have the gRPC server prometheus monitoring. I added a TODO in server/ca/monitoring.go. Will implement it in follow up PRs.

File: security/pkg/server/ca/server.go [hunk 2]
```diff
@@ -60,30 +60,32 @@ func (s *Server) HandleCSR(ctx context.Context, request *pb.CsrRequest) (*pb.Csr
...
+		s.monitoring.AuthnError.Inc()
 		return nil, status.Error(codes.Unauthenticated, "request authenticate failure")
 	}
 
 	csr, err := util.ParsePemEncodedCSR(request.CsrPem)
 	if err != nil {
-		log.Warnf("CSR parsing error (error %v)", err)
-		s.monitoring.CSRParsingError.Inc()
+		log.Warnf("CSR Pem parsing error (error %v)", err)
+		s.monitoring.GetAuthzError("CSR Pem parsing error").Inc()
```
[the remark this answers is not in the document]
Comment (myidpt) [reply security/pkg/server/ca/server.go]: It's actually serving the authz purpose. But I can make the metric explicit.

File: security/pkg/pki/ca/ca.go [hunk 3]
```diff
@@ -171,26 +170,26 @@ func NewIstioCA(opts *IstioCAOptions) (*IstioCA, error) {
 // Sign takes a PEM-encoded CSR and ttl, and returns a signed certificate. If forCA is true,
 // the signed certificate is a CA certificate, otherwise, it is a workload certificate.
 // TODO(myidpt): Add error code to identify the Sign error types.
```
[the remark this answers is not in the document]
Comment (myidpt) [reply security/pkg/pki/ca/ca.go]: Done.

File: security/pkg/server/ca/monitoring.go [hunk 4]
```diff
@@ -18,43 +18,49 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+// TODO: Add gRPC server monitoring.
```
[the remark this answers is not in the document]
Comment (myidpt) [reply security/pkg/server/ca/monitoring.go]: That's a good point. I added the interceptor. It counts the response based on the grpc error code. For example, the metrics look like:
```
grpc_server_handled_total{grpc_code="Aborted",grpc_method="HandleCSR",grpc_service="istio.v1.auth.IstioCAService",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="HandleCSR",grpc_service="istio.v1.auth.IstioCAService",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="HandleCSR",grpc_service="istio.v1.auth.IstioCAService",grpc_type="unary"} 0
```
The only issue is that it does not look deeper into the response message (also it shouldn't). So I think maybe we can keep both grpc interceptor and the customized server monitoring. The former one counts the response code, and the latter one gives insight on the errors.

File: security/pkg/pki/ca/error.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ca
+
+// ErrType defines the enum for CA error types.
```
[the remark this answers is not in the document]
Comment (myidpt) [reply security/pkg/pki/ca/error.go]: Sure.

**Comment by incfly** [comment]
FYI, I think we're now focus on developing on collab-gcp-identity and at some time we need to align the CSR API. This might affect quite bit of things.

**Pull request merged**